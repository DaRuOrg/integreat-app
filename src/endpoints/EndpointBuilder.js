import { isEqual } from 'lodash/lang'

import Endpoint from './Endpoint'
import StateMapperBuilder from './StateMapperBuilder'

/**
 * Helper class to build a {@link Endpoint}
 */
class EndpointBuilder {
  _name
  _url
  _mapper
  _stateMapperBuilder
  _refetchLogic

  /**
   * Creates a new endpoint builder
   * @param {string} name The name of the endpoint to build
   */
  constructor (name) {
    this._name = name
    this._stateMapperBuilder = new StateMapperBuilder(this)

    this._refetchLogic = (urlParams, nextUrlParams) => !isEqual(urlParams, nextUrlParams)
  }

  /**
   * Adds a url to the builder
   * @param url The url
   * @return {EndpointBuilder} The builder itself
   */
  withUrl (url) {
    this._url = url
    return this
  }

  /**
   * Adds a url to the builder
   * @param mapper The mapper
   * @return {EndpointBuilder} The builder itself
   */
  withMapper (mapper) {
    this._mapper = mapper
    return this
  }

  /**
   * Adds a state mapper builder to this builder
   * @return {StateMapperBuilder} The state mapper of this builder
   */
  withStateMapper () {
    return this._stateMapperBuilder
  }

  /**
   * Adds refetch logic to this builder
   * @param refetchLogic The refetch logic
   * @return {EndpointBuilder}  The builder itself
   */
  withRefetchLogic (refetchLogic) {
    this._refetchLogic = refetchLogic
    return this
  }

  /**
   * Checks the data and builds the endpoint
   * @return {Endpoint} The final endpoint
   */
  build () {
    if (!this._name) {
      throw Error('You have to set a name to build an endpoint!')
    }

    if (!this._url) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._mapper) {
      throw Error('You have to set a mapper to build an endpoint!')
    }

    if (!this._url) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._stateMapperBuilder) {
      throw Error('You have to set a state mapper to build an endpoint!')
    }

    if (!this._refetchLogic) {
      throw Error('You have to set a refetch logic to build an endpoint!')
    }

    return new Endpoint(this._name, this._url, this._mapper, this._stateMapperBuilder.build(), this._refetchLogic)
  }
}

export default EndpointBuilder
