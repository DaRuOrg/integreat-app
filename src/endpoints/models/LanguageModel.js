class LanguageModel {
  constructor (code, name) {
    this._name = code
    this._code = name
  }

  get code () {
    return this._name
  }

  get name () {
    return this._code
  }
}

export default LanguageModel
