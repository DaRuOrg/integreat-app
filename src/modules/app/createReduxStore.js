// @flow

import type { Store } from 'redux'
import { applyMiddleware, combineReducers, compose, createStore } from 'redux'
import { createLogger } from 'redux-logger'

import { languageReducer, uiDirectionReducer } from '../i18n/reducers'
import toggleDarkModeReducer from '../theme/reducers'
import Payload from '../endpoint/Payload'

export type ActionType<T> = { type: string, payload: Payload<T> }

// todo: Change type to correct State type,
// https://blog.callstack.io/type-checking-react-and-redux-thunk-with-flow-part-2-206ce5f6e705
const createReduxStore = (initialState: {} = {}): Store<any, any> => {
  /**
   * The middlewares of this app, add additional middlewares here
   */
  const middlewares = []

  if (__DEV__) {
    middlewares.push(createLogger()) // Logs all state changes in console
  }

  const rootReducer = combineReducers({
    uiDirection: uiDirectionReducer,
    language: languageReducer,
    darkMode: toggleDarkModeReducer
  })

  const enhancers = compose(applyMiddleware(...middlewares))

  return createStore(rootReducer, initialState, enhancers)
}

export default createReduxStore
