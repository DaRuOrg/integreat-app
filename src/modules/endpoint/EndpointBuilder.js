import { isEqual } from 'lodash/lang'

import Endpoint from './Endpoint'

/**
 * Helper class to build a {@link Endpoint}
 */
class EndpointBuilder {
  _name
  _routerToUrlMapper
  _mapper
  _responseOverride
  _refetchLogic

  /**
   * Creates a new endpoint builder
   * @param {string} name The name of the endpoint to build
   */
  constructor (name) {
    this._name = name

    this._refetchLogic = (urlParams, nextUrlParams) => !isEqual(urlParams, nextUrlParams)
  }

  /**
   * Adds a router to url mapper to the builder
   * @param routerToUrlMapper The url
   * @return {EndpointBuilder} The builder itself
   */
  withRouterToUrlMapper (routerToUrlMapper) {
    this._routerToUrlMapper = routerToUrlMapper
    return this
  }

  /**
   * Adds a url to the builder
   * @param mapper The mapper
   * @return {EndpointBuilder} The builder itself
   */
  withMapper (mapper) {
    this._mapper = mapper
    return this
  }

  /**
   * Adds refetch logic to this builder
   * @param refetchLogic The refetch logic
   * @return {EndpointBuilder} The builder itself
   */
  withRefetchLogic (refetchLogic) {
    this._refetchLogic = refetchLogic
    return this
  }

  /**
   * Overrides value from the API response. Useful for testing.
   * @param responseOverride {*} The response
   * @return {EndpointBuilder} The builder itself
   */
  withResponseOverride (responseOverride) {
    this._responseOverride = responseOverride
    return this
  }

  /**
   * Checks the data and builds the endpoint
   * @return {Endpoint} The final endpoint
   */
  build () {
    if (!this._name) {
      throw Error('You have to set a name to build an endpoint!')
    }

    if (!this._routerToUrlMapper) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._mapper) {
      throw Error('You have to set a mapper to build an endpoint!')
    }

    if (!this._refetchLogic) {
      throw Error('You have to set a refetch logic to build an endpoint!')
    }

    return new Endpoint(this._name, this._routerToUrlMapper, this._mapper, this._refetchLogic, this._responseOverride)
  }
}

export default EndpointBuilder
