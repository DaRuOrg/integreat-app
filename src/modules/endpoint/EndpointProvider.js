import React from 'react'
import PropTypes from 'prop-types'
import Endpoint from './Endpoint'
import Error from '../common/containers/Error'

class EndpointProvider extends React.Component {
  static propTypes = {
    endpoints: PropTypes.arrayOf(PropTypes.instanceOf(Endpoint)).isRequired,
    children: PropTypes.element.isRequired
  }

  static childContextTypes = {
    getEndpoint: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      getEndpoint: (endpointName) => {
        const endpoint = this.endpoints[endpointName]
        if (!endpoint) {
          throw new Error('The endpoint ' + endpointName + ' was not found. Maybe you didn\'t wrap the ' +
            'withFetcher(...) call in a <EndpointProvider /> or the endpoint is not registered')
        }
        return endpoint
      }
    }
  }

  constructor (props, context) {
    super(props, context)
    this.endpoints = props.endpoints.reduce((accumulator, endpoint) => {
      accumulator[endpoint.stateName] = endpoint
      return accumulator
    }, {})
  }

  render () {
    return React.Children.only(this.props.children)
  }
}

export default EndpointProvider
