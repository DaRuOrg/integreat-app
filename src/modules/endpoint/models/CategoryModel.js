// @flow

type categoryModel = {
  id: number,
  url: string,
  title: string,
  parentId: number,
  parentUrl: string,
  content: string,
  thumbnail: string,
  order: number,
  availableLanguages: any
}

class CategoryModel {
   _id: number
  _url: string
  _title: string
  _parentId: number
  _parentUrl: string
  _content: string
  _thumbnail: string
  _order: number
  _availableLanguages: any

  constructor (categoryModel: categoryModel) {
    this._id = categoryModel.id
    this._url = categoryModel.url
    this._title = categoryModel.title
    this._content = categoryModel.content
    this._parentId = categoryModel.parentId
    this._parentUrl = categoryModel.parentUrl
    this._thumbnail = categoryModel.thumbnail
    this._order = categoryModel.order
    this._availableLanguages = categoryModel.availableLanguages
  }

  get thumbnail (): string {
    return this._thumbnail
  }

  get id (): number {
    return this._id
  }

  get url (): string {
    return this._url
  }

  get title (): string {
    return this._title
  }

  get content (): string {
    return this._content
  }

  get parentId (): number {
    return this._parentId
  }

  get parentUrl (): string {
    return this._parentUrl
  }

  setParentUrl (parentUrl: string) {
    this._parentUrl = parentUrl
  }

  get order (): number {
    return this._order
  }

  get availableLanguages (): any {
    return this._availableLanguages
  }
}

export default CategoryModel
