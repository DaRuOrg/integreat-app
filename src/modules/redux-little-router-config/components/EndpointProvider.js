import React from 'react'
import PropTypes from 'prop-types'
import RouteConfig from '../RouteConfig'

class EndpointProvider extends React.Component {
  static propTypes = {
    routeConfig: PropTypes.instanceOf(RouteConfig).isRequired,
    children: PropTypes.element.isRequired
  }

  static childContextTypes = {
    matchRoute: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      matchRoute: (id) => this.props.routeConfig.matchRoute(id)
    }
  }

  render () {
    return React.Children.only(this.props.children)
  }
}

export default EndpointProvider
