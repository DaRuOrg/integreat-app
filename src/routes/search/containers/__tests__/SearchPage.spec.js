import React from 'react'
import ConnectedSearchPage, { SearchPage } from '../SearchPage'
import LanguageModel from 'modules/endpoint/models/LanguageModel'
import CategoryModel from 'modules/endpoint/models/CategoryModel'
import CategoriesMapModel from 'modules/endpoint/models/CategoriesMapModel'
import { mount, shallow } from 'enzyme'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import EndpointBuilder from 'modules/endpoint/EndpointBuilder'
import EndpointProvider from 'modules/endpoint/EndpointProvider'
import createReduxStore from 'modules/app/createReduxStore'
import createHistory from 'modules/app/createHistory'

describe('SearchPage', () => {
  const location = 'augsburg'

  const languages = [
    new LanguageModel('en', 'English'),
    new LanguageModel('de', 'Deutsch'),
    new LanguageModel('ar', 'Arabic')
  ]

  const categoryModels = [
    new CategoryModel({
      id: 0,
      url: '/augsburg/de',
      title: 'augsburg'
    }),
    new CategoryModel({
      id: 3650,
      url: '/augsburg/de/anlaufstellen',
      title: 'Anlaufstellen zu sonstigen Themen',
      content: '',
      parentId: 0,
      parentUrl: '/augsburg/de',
      order: 75,
      availableLanguages: {
        en: 4361, ar: 4367, fa: 4368
      },
      thumbnail: 'https://cms.integreat-ap…/03/Hotline-150x150.png'
    }),
    new CategoryModel({
      id: 3649,
      url: '/augsburg/de/willkommen',
      title: 'Willkommen',
      content: '',
      parentId: 0,
      parentUrl: '/augsburg/de',
      order: 11,
      availableLanguages: {
        en: 4804, ar: 4819, fa: 4827
      },
      thumbnail: 'https://cms.integreat-ap…03/Beratung-150x150.png'
    }),
    new CategoryModel({
      id: 35,
      url: '/augsburg/de/willkommen/willkommen-in-augsburg',
      title: 'Willkommen in Augsburg',
      content: '<p>Willkommen in Augsbur…er Stadt Augsburg</p>\n',
      parentId: 3649,
      parentUrl: '/augsburg/de/willkommen',
      order: 1,
      availableLanguages: {
        en: '390',
        de: '711',
        ar: '397'
      },
      thumbnail: 'https://cms.integreat-ap…09/heart295-150x150.png'
    })
  ]

  const categories = new CategoriesMapModel(categoryModels)

  test('should match snapshot', () => {
    const wrapper = shallow(<SearchPage location={location}
                                        languages={languages}
                                        categories={categories}
                                        setLanguageChangeUrls={() => {}} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should dispatch once in componentDidMount', () => {
    const mockSetLanguageChangeUrls = jest.fn()

    const searchPage = shallow(<SearchPage location={location}
                                           languages={languages}
                                           categories={categories}
                                           setLanguageChangeUrls={mockSetLanguageChangeUrls} />
    ).instance()

    expect(mockSetLanguageChangeUrls.mock.calls).toHaveLength(1)
    expect(mockSetLanguageChangeUrls).toBeCalledWith(searchPage.mapLanguageToUrl, languages)
  })

  test('should mapLanguageToUrl correctly', () => {
    const searchPage = shallow(
      <SearchPage location={location}
                  languages={languages}
                  categories={categories}
                  setLanguageChangeUrls={() => {}} />
    ).instance()
    expect(searchPage.mapLanguageToUrl('en')).toBe('/augsburg/en/search')
  })

  test('should filter correctly', () => {
    const mockStore = configureMockStore()
    const store = mockStore({router: {}})

    const tree = mount(
      <Provider store={store}>
        <SearchPage location={location}
                    languages={languages}
                    categories={categories}
                    setLanguageChangeUrls={() => {}} />
      </Provider>
    )
    const searchPage = tree.find(SearchPage).instance()
    const searchInputProps = tree.find('SearchInput').props()

    expect(searchPage.findCategories()).toHaveLength(categories.toArray().length)

    searchInputProps.onFilterTextChange('Does not exist!')
    expect(searchPage.findCategories()).toHaveLength(0)

    searchInputProps.onFilterTextChange(categoryModels[1].title)
    expect(searchPage.findCategories()).toHaveLength(1)
  })

  describe('connect()', () => {
    const categoriesEndpoint = new EndpointBuilder('categories')
      .withStateToUrlMapper(() => 'https://weird-endpoint/api.json')
      .withMapper(json => json)
      .withResponseOverride(categories)
      .build()

    const languagesEndpoint = new EndpointBuilder('languages')
      .withStateToUrlMapper(() => 'https://weird-endpoint/api.json')
      .withMapper(json => json)
      .withResponseOverride(languages)
      .build()

    test('should map state to props', () => {
      const store = createReduxStore(createHistory, {
        router: {params: {location: location}}
      })

      const tree = mount(
        <Provider store={store}>
          <EndpointProvider endpoints={[categoriesEndpoint, languagesEndpoint]}>
            <ConnectedSearchPage />
          </EndpointProvider>
        </Provider>
      )

      const categoriesPageProps = tree.find(SearchPage).props()

      expect(categoriesPageProps).toEqual({
        location,
        categories,
        languages,
        setLanguageChangeUrls: expect.any(Function)
      })
    })

    test('should map dispatch to props', () => {
      const store = createReduxStore(createHistory, {
        router: {params: {location: location}}
      })

      expect(store.getState().languageChangeUrls).toEqual({})

      mount(
        <Provider store={store}>
          <EndpointProvider endpoints={[categoriesEndpoint, languagesEndpoint]}>
            <ConnectedSearchPage />
          </EndpointProvider>
        </Provider>
      )

      expect(store.getState().languageChangeUrls).toEqual({
        en: '/augsburg/en/search',
        de: '/augsburg/de/search',
        ar: '/augsburg/ar/search'
      })
    })
  })
})
