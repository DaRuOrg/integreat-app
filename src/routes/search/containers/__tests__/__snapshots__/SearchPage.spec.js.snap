// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage should match snapshot 1`] = `
<div>
  <Component
    className="searchSpacing"
    filterText=""
    onFilterTextChange={[Function]}
  />
  <CategoryList
    categories={
      Array [
        CategoryModel {
          "_availableLanguages": Object {},
          "_content": "",
          "_id": 0,
          "_order": 0,
          "_parentId": -1,
          "_parentUrl": null,
          "_thumbnail": null,
          "_title": "augsburg",
          "_url": "/augsburg/de",
        },
        CategoryModel {
          "_availableLanguages": Object {
            "ar": 4367,
            "en": 4361,
            "fa": 4368,
          },
          "_content": "",
          "_id": 3650,
          "_order": 75,
          "_parentId": 0,
          "_parentUrl": "/augsburg/de",
          "_thumbnail": "https://cms.integreat-ap…/03/Hotline-150x150.png",
          "_title": "Anlaufstellen zu sonstigen Themen",
          "_url": "/augsburg/de/anlaufstellen",
        },
        CategoryModel {
          "_availableLanguages": Object {
            "ar": 4819,
            "en": 4804,
            "fa": 4827,
          },
          "_content": "",
          "_id": 3649,
          "_order": 11,
          "_parentId": 0,
          "_parentUrl": "/augsburg/de",
          "_thumbnail": "https://cms.integreat-ap…03/Beratung-150x150.png",
          "_title": "Willkommen",
          "_url": "/augsburg/de/willkommen",
        },
        CategoryModel {
          "_availableLanguages": Object {
            "ar": "397",
            "de": "711",
            "en": "390",
          },
          "_content": "<p>Willkommen in Augsbur…er Stadt Augsburg</p>
",
          "_id": 35,
          "_order": 1,
          "_parentId": 3649,
          "_parentUrl": "/augsburg/de/willkommen",
          "_thumbnail": "https://cms.integreat-ap…09/heart295-150x150.png",
          "_title": "Willkommen in Augsburg",
          "_url": "/augsburg/de/willkommen/willkommen-in-augsburg",
        },
      ]
    }
  />
</div>
`;
